govcd/vm.go

```diff
@@ -850,3 +851,186 @@ func (vm *VM) SetGuestCustomizationSection(guestCustomizationSection *types.Gues
 
 	return vm.GetGuestCustomizationSection()
 }
+
+// AddInternalDisk creates disk type *types.DiskSettings to the VM.
+// Returns new disk ID and error.
+// Runs synchronously, VM is ready for another operation after this function returns.
+func (vm *VM) AddInternalDisk(diskData *types.DiskSettings) (string, error) {
+	if vm.VM.HREF == "" {
+		return "", fmt.Errorf("cannot add internal disks - VM HREF is unset")
+	}
+
+	err := vm.validateInternalDiskInput(diskData)
+	if err != nil {
+		return "", err
+	}
+
+	diskSettings := vm.VM.VmSpecSection.DiskSection.DiskSettings
+	if diskSettings == nil {
+		diskSettings = []*types.DiskSettings{}
+	}
+
+	diskSettings = append(diskSettings, diskData)
+	vmSpecSection := vm.VM.VmSpecSection
+	vmSpecSection.DiskSection.DiskSettings = diskSettings
+
+	vmSpecSection, err = vm.UpdateInternalDisks(vmSpecSection)
+	if err != nil {
+		return "", err
+	}
+
+	for _, diskSetting := range vmSpecSection.DiskSection.DiskSettings {
+		if diskSetting.AdapterType == diskData.AdapterType &&
+			diskSetting.BusNumber == diskData.BusNumber &&
+			diskSetting.UnitNumber == diskData.UnitNumber {
+			return diskSetting.DiskId, nil
+		}
+	}
+
+	return "", fmt.Errorf("created disk wasn't in list of returned VM internal disks")
+}
+
+func (vm *VM) validateInternalDiskInput(diskData *types.DiskSettings) error {
+	if diskData.AdapterType == "" {
+		return fmt.Errorf("disk settings missing required field: adapter type")
+	}
+
+	if diskData.BusNumber < 0 {
```

Comment:
I think this validation cannot be hit because the type is `int` and an empty value will be 0. Same for UnitNumber and SizeMb below. Go itself will "default" it to 0.

Reply:
This is govcd, not only terraform can call it. So user could provide for e.g. `-1`

Reply:
True, but my main concern is probably the values being `int` instead of `*int` as if user didn't set anything - we will get 0. Up to you if it looks to be risky.

Reply:
it wouldn't be correct to make *int from vCD API side

Reply:
Agreed.